storage/src/main/java/tech/pegasys/teku/storage/server/VersionedHashDBSourceFactory.java

```diff
@@ -0,0 +1,89 @@
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.storage.server;
+
+import java.util.function.Function;
+import org.apache.commons.lang3.tuple.Pair;
+import tech.pegasys.teku.infrastructure.events.EventChannels;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.spec.SpecMilestone;
+import tech.pegasys.teku.spec.config.SpecConfigFulu;
+import tech.pegasys.teku.spec.datastructures.blobs.versions.deneb.BlobSidecar;
+import tech.pegasys.teku.spec.datastructures.blobs.versions.fulu.DataColumnSidecar;
+import tech.pegasys.teku.spec.logic.versions.deneb.helpers.MiscHelpersDeneb;
+import tech.pegasys.teku.spec.logic.versions.deneb.types.VersionedHash;
+import tech.pegasys.teku.statetransition.api.CustodyGroupCountChannel;
+import tech.pegasys.teku.storage.server.kvstore.dataaccess.KvStoreCombinedDao;
+import tech.pegasys.teku.storage.server.kvstore.dataaccess.VersionedHashDBSource;
+
+public class VersionedHashDBSourceFactory {
+
+  final Function<BlobSidecar, VersionedHash> blobSidecarToVersionedHash;
+  final Function<Pair<DataColumnSidecar, UInt64>, VersionedHash> dataColumnSidecarToVersionedHash;
+  final EventChannels eventChannels;
+  final Spec spec;
+
+  public VersionedHashDBSourceFactory(final Spec spec, final EventChannels eventChannels) {
+    this.spec = spec;
+    this.eventChannels = eventChannels;
+    if (spec.isMilestoneSupported(SpecMilestone.DENEB)) {
+      this.blobSidecarToVersionedHash =
+          blobSidecar ->
+              MiscHelpersDeneb.required(spec.forMilestone(SpecMilestone.DENEB).miscHelpers())
+                  .kzgCommitmentToVersionedHash(blobSidecar.getKZGCommitment());
+    } else {
+      this.blobSidecarToVersionedHash = blobSidecar -> VersionedHash.EMPTY;
+    }
+    if (spec.isMilestoneSupported(SpecMilestone.FULU)) {
+      this.dataColumnSidecarToVersionedHash =
+          pair ->
+              MiscHelpersDeneb.required(spec.forMilestone(SpecMilestone.DENEB).miscHelpers())
+                  .kzgCommitmentToVersionedHash(
+                      pair.getKey()
+                          .getSszKZGCommitments()
+                          .get(pair.getValue().intValue())
+                          .getKZGCommitment());
+    } else {
+      this.dataColumnSidecarToVersionedHash = pair -> VersionedHash.EMPTY;
+    }
+  }
+
+  public VersionedHashDBSource createVersionedHashDBSource(final KvStoreCombinedDao dao) {
+    final VersionedHashDBSource versionedHashDBSource =
+        new VersionedHashDBSource(
+            dao, blobSidecarToVersionedHash, dataColumnSidecarToVersionedHash, spec);
+    eventChannels.subscribe(
+        CustodyGroupCountChannel.class,
+        new CustodyGroupCountChannel() {
+          @Override
+          public void onGroupCountUpdate(int custodyGroupCount, int samplingGroupCount) {
```

Comment:
a question here: do we receive this event here quick enough even when the node is a supernode due to number of ETH attached and not just due to explicit custody param?

I'm worried if we are able to set the flag quick enough and avoid loosing versioned hashed

Reply:
Don't we save how much we custody in a db var, so that it can go up only as per spec? I just remember this, maybe i'm wrong

Reply:
Yeah, we save, but we cannot put db read in constructor (or can?). I mean it's slow